File: backend/src/tripl/api/v1/alerting.py
import uuid
from datetime import datetime

from fastapi import APIRouter, HTTPException, Query
from sqlalchemy import select

from tripl.api.deps import EditorUserDep, SessionDep
from tripl.models.alert_delivery import AlertDeliveryStatus
from tripl.models.alert_destination import AlertDestinationType
from tripl.models.alert_rule import AlertRule
from tripl.models.anomaly_scope_override import RATCHET_SIGMA_CAP
from tripl.models.domain_enums import AlertInboxStatus
from tripl.schemas.alerting import (
    AlertDeliveryDetailResponse,
    AlertDeliveryListResponse,
    AlertDestinationCreate,
    AlertDestinationResponse,
    AlertDestinationTestResponse,
    AlertDestinationUpdate,
    AlertInboxActionRequest,
    AlertInboxActionResponse,
    AlertInboxBulkActionRequest,
    AlertInboxBulkActionResponse,
    AlertInboxGroupResponse,
    AlertInboxListResponse,
    AlertRuleCreate,
    AlertRuleResponse,
    AlertRuleSimulateResponse,
    AlertRuleUpdate,
    MonitorDetailResponse,
    MonitorMuteRequest,
    MonitorsSummaryResponse,
)
from tripl.services import alerting_service, audit_service

router = APIRouter(prefix="/projects/{slug}", tags=["alerting"])


@router.get("/alert-destinations", response_model=list[AlertDestinationResponse])
async def list_alert_destinations(session: SessionDep, slug: str) -> list[AlertDestinationResponse]:
    return await alerting_service.list_destinations(session, slug)


@router.get("/monitors-summary", response_model=MonitorsSummaryResponse)
async def get_monitors_summary(session: SessionDep, slug: str) -> MonitorsSummaryResponse:
    return await alerting_service.get_monitors_summary(session, slug)


@router.post("/alert-destinations", response_model=AlertDestinationResponse, status_code=201)
async def create_alert_destination(
    session: SessionDep,
    slug: str,
    data: AlertDestinationCreate,
    current_user: EditorUserDep,
) -> AlertDestinationResponse:
    dest = await alerting_service.create_destination(session, slug, data)
    await audit_service.record(
        session,
        user=current_user,
        action="alert_destination.create",
        target_type="alert_destination",
        target_id=dest.id,
        target_name=dest.name,
        project_slug=slug,
        payload=data.model_dump(),
    )
    return dest


@router.get("/alert-destinations/{destination_id}", response_model=AlertDestinationResponse)
async def get_alert_destination(
    session: SessionDep, slug: str, destination_id: uuid.UUID
) -> AlertDestinationResponse:
    return await alerting_service.get_destination(session, slug, destination_id)


@router.patch("/alert-destinations/{destination_id}", response_model=AlertDestinationResponse)
async def update_alert_destination(
    session: SessionDep,
    slug: str,
    destination_id: uuid.UUID,
    data: AlertDestinationUpdate,
    current_user: EditorUserDep,
) -> AlertDestinationResponse:
    dest = await alerting_service.update_destination(session, slug, destination_id, data)
    await audit_service.record(
        session,
        user=current_user,
        action="alert_destination.update",
        target_type="alert_destination",
        target_id=dest.id,
        target_name=dest.name,
        project_slug=slug,
        payload=data.model_dump(exclude_unset=True),
    )
    return dest


@router.delete("/alert-destinations/{destination_id}", status_code=204)
async def delete_alert_destination(
    session: SessionDep,
    slug: str,
    destination_id: uuid.UUID,
    current_user: EditorUserDep,
) -> None:
    # The name comes back from the delete itself, which already loaded the row.
    # Asking the service for the destination first meant
    # ``get_destination_response`` — the four delete-impact aggregates of
    # ``load_destination_health`` — to read one string off a row that is about to
    # be deleted.
    name = await alerting_service.delete_destination(session, slug, destination_id)
    await audit_service.record(
        session,
        user=current_user,
        action="alert_destination.delete",
        target_type="alert_destination",
        target_id=destination_id,
        target_name=name,
        project_slug=slug,
    )


@router.post(
    "/alert-destinations/{destination_id}/test",
    response_model=AlertDestinationTestResponse,
)
async def test_alert_destination(
    session: SessionDep,
    slug: str,
    destination_id: uuid.UUID,
    current_user: EditorUserDep,
) -> AlertDestinationTestResponse:
    """Send one fixed test message through this destination's real channel.

    Editor-only: it puts a message in somebody's Slack/Telegram/inbox and, for a
    tracker destination, opens a ticket. Always 200 — a channel refusal is the
    answer the caller asked for, not a server fault (see
    ``AlertDestinationTestResponse``).

    Recorded in the audit log rather than as an AlertDelivery, so the Delivery
    log keeps meaning "an alert fired"; see ``services/_alerting_test_send``.
    """
    # The name comes back with the result, read off the load the send already
    # did and taken BEFORE the network call — so the audit entry still names the
    # destination the operator acted on even if the row changes underneath. It
    # used to be fetched by a second call to ``get_destination``, which is
    # ``get_destination_response``: four delete-impact aggregates and a second
    # load of the same row, in front of a request that blocks for up to 10s.
    outcome = await alerting_service.send_destination_test(session, slug, destination_id)
    result = outcome.response
    await audit_service.record(
        session,
        user=current_user,
        action="alert_destination.test",
        target_type="alert_destination",
        target_id=destination_id,
        target_name=outcome.destination_name,
        project_slug=slug,
        payload={"ok": result.ok, "error": result.error},
    )
    return result


@router.post(
    "/alert-destinations/{destination_id}/rules",
    response_model=AlertRuleResponse,
    status_code=201,
)
async def create_alert_rule(
    session: SessionDep,
    slug: str,
    destination_id: uuid.UUID,
    data: AlertRuleCreate,
    current_user: EditorUserDep,
) -> AlertRuleResponse:
    rule = await alerting_service.create_rule(session, slug, destination_id, data)
    await audit_service.record(
        session,
        user=current_user,
        action="alert_rule.create",
        target_type="alert_rule",
        target_id=rule.id,
        target_name=rule.name,
        project_slug=slug,
        payload=data.model_dump(),
    )
    return rule


@router.patch(
    "/alert-destinations/{destination_id}/rules/{rule_id}",
    response_model=AlertRuleResponse,
)
async def update_alert_rule(
    session: SessionDep,
    slug: str,
    destination_id: uuid.UUID,
    rule_id: uuid.UUID,
    data: AlertRuleUpdate,
    current_user: EditorUserDep,
) -> AlertRuleResponse:
    rule = await alerting_service.update_rule(session, slug, destination_id, rule_id, data)
    await audit_service.record(
        session,
        user=current_user,
        action="alert_rule.update",
        target_type="alert_rule",
        target_id=rule.id,
        target_name=rule.name,
        project_slug=slug,
        payload=data.model_dump(exclude_unset=True),
    )
    return rule


@router.delete("/alert-destinations/{destination_id}/rules/{rule_id}", status_code=204)
async def delete_alert_rule(
    session: SessionDep,
    slug: str,
    destination_id: uuid.UUID,
    rule_id: uuid.UUID,
    current_user: EditorUserDep,
) -> None:
    rule = await session.scalar(
        select(AlertRule).where(AlertRule.id == rule_id, AlertRule.destination_id == destination_id)
    )
    if rule is None:
        raise HTTPException(status_code=404, detail="Alert rule not found")
    name = rule.name
    await alerting_service.delete_rule(session, slug, destination_id, rule_id)
    await audit_service.record(
        session,
        user=current_user,
        action="alert_rule.delete",
        target_type="alert_rule",
        target_id=rule_id,
        target_name=name,
        project_slug=slug,
    )


@router.post(
    "/alert-destinations/{destination_id}/rules/{rule_id}/simulate",
    response_model=AlertRuleSimulateResponse,
)
async def simulate_alert_rule(
    session: SessionDep,
    slug: str,
    destination_id: uuid.UUID,
    rule_id: uuid.UUID,
    days: int = Query(7, ge=1, le=90),
    cooldown_minutes_override: int | None = Query(None, ge=0, le=10080),
    # The three what-if knobs beside the cooldown one. Omit any of them and the
    # replay uses the rule's saved value; the reply reports both, as
    # ``*_used``/``*_saved``. Without these, asking "would min % 300 cut these
    # incidents" meant saving 300 onto a rule that is live-routing to a real
    # channel (tripl-oxkt.17 part 3).
    #
    # ``sigma_threshold_override`` is the DETECTOR's sensitivity, not a rule
    # field: it re-reads the recorded anomalies, so it can only ever narrow the
    # set — see ``AlertRuleSimulateResponse``. Bounded above by the same cap the
    # false-positive ratchet respects.
    min_percent_delta_override: float | None = Query(None, ge=0),
    min_expected_count_override: float | None = Query(None, ge=0),
    sigma_threshold_override: float | None = Query(None, gt=0, le=RATCHET_SIGMA_CAP),
) -> AlertRuleSimulateResponse:
    return await alerting_service.simulate_rule(
        session,
        slug,
        destination_id,
        rule_id,
        days,
        cooldown_minutes_override=cooldown_minutes_override,
        min_percent_delta_override=min_percent_delta_override,
        min_expected_count_override=min_expected_count_override,
        sigma_threshold_override=sigma_threshold_override,
    )


@router.get("/alert-deliveries", response_model=AlertDeliveryListResponse)
async def list_alert_deliveries(
    session: SessionDep,
    slug: str,
    # AlertDeliveryStatus / AlertDestinationType (not str): both columns are
    # native Postgres enums, so a garbage value used to travel all the way to
    # the driver and blow up as a 500 ("not among the defined enum values").
    # FastAPI now 422s it at the edge, like GET /events?status= already did
    # (tripl-57g0).
    status: AlertDeliveryStatus | None = None,
    channel: AlertDestinationType | None = None,
    destination_id: uuid.UUID | None = None,
    rule_id: uuid.UUID | None = None,
    scan_config_id: uuid.UUID | None = None,
    # The page nests deliveries under their incident, so it fetches them one
    # incident at a time; `ungrouped` reaches the rows that have no incident,
    # which no value of `correlation_group_id` can select (tripl-pq97).
    correlation_group_id: uuid.UUID | None = None,
    ungrouped: bool = False,
    date_from: datetime | None = None,
    date_to: datetime | None = None,
    offset: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=200),
) -> AlertDeliveryListResponse:
    # Contradictory by construction: a delivery with an item in the group
    # necessarily has an item WITH a group, which `ungrouped` excludes, so the
    # pair can only ever return zero rows. Rejecting it is the difference between
    # "you asked for something impossible" and "there is nothing here" — which
    # read identically to a caller holding an empty list.
    if correlation_group_id is not None and ungrouped:
        raise HTTPException(
            status_code=422,
            detail="correlation_group_id and ungrouped are mutually exclusive",
        )
    return await alerting_service.list_deliveries(
        session,
        slug,
        status=status,
        channel=channel,
        destination_id=destination_id,
        rule_id=rule_id,
        scan_config_id=scan_config_id,
        correlation_group_id=correlation_group_id,
        ungrouped=ungrouped,
        date_from=date_from,
        date_to=date_to,
        offset=offset,
        limit=limit,
    )


@router.get("/alert-deliveries/{delivery_id}", response_model=AlertDeliveryDetailResponse)
async def get_alert_delivery(
    session: SessionDep, slug: str, delivery_id: uuid.UUID
) -> AlertDeliveryDetailResponse:
    return await alerting_service.get_delivery(session, slug, delivery_id)


@router.post(
    "/alert-deliveries/{delivery_id}/retry",
    response_model=AlertDeliveryDetailResponse,
)
async def retry_alert_delivery(
    session: SessionDep,
    slug: str,
    delivery_id: uuid.UUID,
    current_user: EditorUserDep,
) -> AlertDeliveryDetailResponse:
    delivery = await alerting_service.retry_delivery(session, slug, delivery_id)
    await audit_service.record(
        session,
        user=current_user,
        action="alert_delivery.retry",
        target_type="alert_delivery",
        target_id=delivery.id,
        target_name=delivery.rule_name,
        project_slug=slug,
    )
    return delivery


@router.get("/monitors/{rule_id}", response_model=MonitorDetailResponse)
async def get_monitor(session: SessionDep, slug: str, rule_id: uuid.UUID) -> MonitorDetailResponse:
    return await alerting_service.get_monitor(session, slug, rule_id)


@router.post("/monitors/{rule_id}/mute", response_model=MonitorDetailResponse)
async def mute_monitor(
    session: SessionDep,
    slug: str,
    rule_id: uuid.UUID,
    data: MonitorMuteRequest,
    current_user: EditorUserDep,
) -> MonitorDetailResponse:
    monitor = await alerting_service.mute_monitor(session, slug, rule_id, data.muted_until)
    await audit_service.record(
        session,
        user=current_user,
        action="alert_rule.mute",
        target_type="alert_rule",
        target_id=monitor.rule_id,
        target_name=monitor.rule_name,
        project_slug=slug,
        payload=data.model_dump(mode="json"),
    )
    return monitor


@router.post("/monitors/{rule_id}/unmute", response_model=MonitorDetailResponse)
async def unmute_monitor(
    session: SessionDep,
    slug: str,
    rule_id: uuid.UUID,
    current_user: EditorUserDep,
) -> MonitorDetailResponse:
    monitor = await alerting_service.unmute_monitor(session, slug, rule_id)
    await audit_service.record(
        session,
        user=current_user,
        action="alert_rule.unmute",
        target_type="alert_rule",
        target_id=monitor.rule_id,
        target_name=monitor.rule_name,
        project_slug=slug,
    )
    return monitor


@router.get("/alert-inbox", response_model=AlertInboxListResponse)
async def list_alert_inbox(
    session: SessionDep,
    slug: str,
    # Inbox status is filtered in Python against the derived group status, so a
    # typo never reached the driver — it silently matched nothing and reported
    # "no incidents", which is worse than a crash. AlertInboxStatus makes the
    # typo a 422 (tripl-57g0).
    status: AlertInboxStatus | None = None,
    offset: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=200),
) -> AlertInboxListResponse:
    return await alerting_service.list_alert_inbox(
        session,
        slug,
        status=status,
        offset=offset,
        limit=limit,
    )


# Registered BEFORE the /{correlation_group_id} routes below, and it must STAY
# there. "bulk-actions" is a literal segment competing with a path parameter that
# FastAPI coerces to a UUID, so whichever route is registered first wins the
# match: registered after, any route on the /{correlation_group_id} template that
# shares this one's method would swallow the request and 422 it as a malformed
# UUID — a rejection of a URL that is spelled exactly right, and one that reads
# like a client bug. Today only the POST .../actions and GET /{id} routes sit on
# that template, so nothing collides on POST yet; the ordering is what keeps that
# true when the next route is added, since nothing else in the file would stop
# it. The variables router has the identical hazard with its literal "drifts"
# segment and solves it the same way (api/v1/variables.py).
@router.post("/alert-inbox/bulk-actions", response_model=AlertInboxBulkActionResponse)
async def apply_alert_inbox_bulk_action(
    session: SessionDep,
    slug: str,
    data: AlertInboxBulkActionRequest,
    current_user: EditorUserDep,
) -> AlertInboxBulkActionResponse:
    """Apply one triage decision to a selection of incidents (tripl-gpfr).

    A shortcut for N clicks on the sibling single-incident route, not a new kind
    of object: the decision is copied into each incident's own state. The service
    validates every id before mutating anything and commits once, so this either
    applies to the whole selection or to none of it.
    """
    result = await alerting_service.apply_alert_inbox_bulk_action(
        session,
        slug,
        data,
        current_user.id,
    )
    # One audit row PER GROUP, reusing the single route's own
    # ``alert_inbox.{action}`` names and its ``alert_correlation_group`` target
    # type, so a bulk mute is searchable by exactly the same query as a
    # hand-clicked one. NOT one batch row with ``target_id=None``:
    # ``audit_log.target_id`` holds a single UUID, and an audit trail that cannot
    # say WHICH incident was muted is worse than no trail at all. ``batch_id`` in
    # the payload is what re-joins the N rows into the one click that wrote them.
    #
    # The id list is excluded from the per-row payload: it would repeat the same
    # (up to 200) ids in all 200 rows, and ``target_id`` already names the one
    # incident this row is about. ``batch_size`` keeps the "how big was that
    # click" question answerable from any single row.
    acted_group_ids = alerting_service.dedupe_correlation_group_ids(data.correlation_group_ids)
    action_payload = data.model_dump(exclude={"correlation_group_ids"})
    for group_id in acted_group_ids:
        await audit_service.record(
            session,
            user=current_user,
            action=f"alert_inbox.{data.action}",
            target_type="alert_correlation_group",
            target_id=group_id,
            target_name=str(group_id),
            project_slug=slug,
            payload={
                **action_payload,
                "batch_id": result.batch_id,
                "batch_size": len(acted_group_ids),
            },
            # One transaction for the whole batch. ``record`` commits per call by
            # default, which for 200 incidents meant 200 separate transactions
            # after the state change was already durable — a failure partway
            # through left incidents silenced with only a partial record of who
            # silenced them, and the response still said 200 OK (tripl-gpfr).
            commit=False,
        )
    await session.commit()
    return result


@router.get("/alert-inbox/{correlation_group_id}", response_model=AlertInboxGroupResponse)
async def get_alert_inbox_group(
    session: SessionDep,
    slug: str,
    correlation_group_id: uuid.UUID,
) -> AlertInboxGroupResponse:
    """Resolve one incident by id, ignoring the list's lookback window.

    Alert messages deep-link the incident they describe, and the reader opens
    them late; before this route the link landed on a page of 20 unrelated
    incidents with no explanation (tripl-oxkt.7).
    """
    return await alerting_service.get_alert_inbox_group(session, slug, correlation_group_id)


@router.post("/alert-inbox/{correlation_group_id}/actions", response_model=AlertInboxActionResponse)
async def apply_alert_inbox_action(
    session: SessionDep,
    slug: str,
    correlation_group_id: uuid.UUID,
    data: AlertInboxActionRequest,
    current_user: EditorUserDep,
) -> AlertInboxActionResponse:
    result = await alerting_service.apply_alert_inbox_action(
        session,
        slug,
        correlation_group_id,
        data,
        current_user.id,
    )
    await audit_service.record(
        session,
        user=current_user,
        action=f"alert_inbox.{data.action}",
        target_type="alert_correlation_group",
        target_id=correlation_group_id,
        target_name=str(correlation_group_id),
        project_slug=slug,
        payload=data.model_dump(),
    )
    return result

File: frontend/src/api/alerting.ts
import { api } from './client'
import type {
  AlertDeliveryDetail,
  AlertInboxAction,
  AlertInboxActionResponse,
  AlertInboxBulkAction,
  AlertInboxBulkActionResponse,
  AlertInboxGroup,
  AlertInboxListResponse,
  AlertInboxStatus,
  AlertDeliveryListResponse,
  AlertDestination,
  AlertDestinationTestResponse,
  AlertRule,
  AlertRuleFilterPayload,
  AlertRuleSimulateResponse,
  MonitorDetail,
  MonitorsSummaryResponse,
} from '../types'

/**
 * The largest selection {@link alertingApi.applyInboxBulkAction} may send
 * (tripl-gpfr).
 *
 * The server's half of this number is `MAX_BULK_INBOX_ACTION_GROUPS` in
 * backend/src/tripl/schemas/alerting.py, where it is a `max_length` on the id
 * list: a longer list is a 422 that never touches the database. This is the ONE
 * place the frontend spells it, so the bar that stops an operator at the cap and
 * the request that would be refused by it cannot drift apart — a second literal
 * `200` typed into a component is a cap that silently stops matching the server
 * the day the server's changes.
 *
 * It lives here, beside the client function bound by it, rather than in
 * `pages/alerting/constants.ts`: that module holds this page's form options and
 * message templates, none of which the server constrains, while this is a wire
 * contract and belongs with the wire. `VARIABLES_PAGE_LIMIT` in api/variables.ts
 * is the same shape of constant in the same place, for the same reason.
 *
 * Reachable in practice, which is why the UI must not leave it to the 422: the
 * inbox is an ACCUMULATING infinite list at 50 rows per page, so four "Load
 * more" clicks put 200 tickable rows on screen and the fifth puts 250 there.
 */
export const MAX_BULK_INBOX_ACTION_GROUPS = 200

/**
 * The longest note either inbox route will store (tripl-gwrd).
 *
 * The server's half is `max_length=2000` on `AlertInboxActionRequest.note` AND
 * on `AlertInboxBulkActionRequest.note` (schemas/alerting.py) — already written
 * twice there, and a third time as a bare `maxLength={2000}` on the incident
 * card's editor. The bulk bar's editor would have made four, so it is named once
 * here for the reason the cap above is: a wire contract, beside the wire.
 *
 * Unlike the bulk cap, this one is NOT enforced by refusing the request. Both
 * editors hand it to `maxLength`, which stops accepting keystrokes SILENTLY —
 * so the only way an operator learns the limit exists is by being told before
 * they reach it (see `noteBudgetLabel`).
 */
export const MAX_INBOX_NOTE_LENGTH = 2000

export const alertingApi = {
  listDestinations: (slug: string) =>
    api.get<AlertDestination[]>(`/projects/${slug}/alert-destinations`),

  getMonitorsSummary: (slug: string) =>
    api.get<MonitorsSummaryResponse>(`/projects/${slug}/monitors-summary`),

  createDestination: (
    slug: string,
    data: {
      type: 'slack' | 'telegram' | 'webhook' | 'email' | 'jira' | 'linear'
      name: string
      enabled?: boolean
      webhook_url?: string | null
      bot_token?: string | null
      chat_id?: string | null
      target_url?: string | null
      webhook_header_name?: string | null
      webhook_header_value?: string | null
      email_recipients?: string | null
      email_from_address?: string | null
      email_subject_template?: string | null
      jira_base_url?: string | null
      jira_auth_email?: string | null
      jira_api_token?: string | null
      jira_project_key?: string | null
      jira_issue_type?: string | null
      linear_api_key?: string | null
      linear_team_id?: string | null
      linear_state_id?: string | null
      linear_label_ids?: string | null
    },
  ) => api.post<AlertDestination>(`/projects/${slug}/alert-destinations`, data),

  updateDestination: (
    slug: string,
    destinationId: string,
    data: {
      name?: string
      enabled?: boolean
      webhook_url?: string | null
      bot_token?: string | null
      chat_id?: string | null
      target_url?: string | null
      webhook_header_name?: string | null
      webhook_header_value?: string | null
      email_recipients?: string | null
      email_from_address?: string | null
      email_subject_template?: string | null
      jira_base_url?: string | null
      jira_auth_email?: string | null
      jira_api_token?: string | null
      jira_project_key?: string | null
      jira_issue_type?: string | null
      linear_api_key?: string | null
      linear_team_id?: string | null
      linear_state_id?: string | null
      linear_label_ids?: string | null
    },
  ) => api.patch<AlertDestination>(`/projects/${slug}/alert-destinations/${destinationId}`, data),

  deleteDestination: (slug: string, destinationId: string) =>
    api.del(`/projects/${slug}/alert-destinations/${destinationId}`),

  /**
   * Send a probe through this destination's real channel. Resolves 200 even
   * when the channel refuses — inspect `ok`/`error` rather than catching, since
   * a rejected token IS the answer the button was pressed for (tripl-oxkt.17).
   */
  testDestination: (slug: string, destinationId: string) =>
    api.post<AlertDestinationTestResponse>(
      `/projects/${slug}/alert-destinations/${destinationId}/test`,
      undefined,
    ),

  createRule: (
    slug: string,
    destinationId: string,
    data: {
      name: string
      enabled?: boolean
      include_project_total?: boolean
      include_event_types?: boolean
      include_events?: boolean
      include_schema_drifts?: boolean
      include_distribution_drifts?: boolean
      include_release_regressions?: boolean
      include_variable_value_drifts?: boolean
      // Accepted by the API and carried on AlertRule, but was missing from both
      // payloads — a form could show the metric toggle and never save it.
      include_metrics?: boolean
      notify_on_spike?: boolean
      notify_on_drop?: boolean
      ai_explanation_enabled?: boolean
      min_percent_delta?: number
      min_absolute_delta?: number
      min_expected_count?: number
      cooldown_minutes?: number
      message_template?: string | null
      items_template?: string | null
      message_format?: 'plain' | 'slack_mrkdwn' | 'telegram_html' | 'telegram_markdownv2'
      filters?: AlertRuleFilterPayload[]
    },
  ) => api.post<AlertRule>(`/projects/${slug}/alert-destinations/${destinationId}/rules`, data),

  updateRule: (
    slug: string,
    destinationId: string,
    ruleId: string,
    data: {
      name?: string
      enabled?: boolean
      include_project_total?: boolean
      include_event_types?: boolean
      include_events?: boolean
      include_schema_drifts?: boolean
      include_distribution_drifts?: boolean
      include_release_regressions?: boolean
      include_variable_value_drifts?: boolean
      // Accepted by the API and carried on AlertRule, but was missing from both
      // payloads — a form could show the metric toggle and never save it.
      include_metrics?: boolean
      notify_on_spike?: boolean
      notify_on_drop?: boolean
      ai_explanation_enabled?: boolean
      min_percent_delta?: number
      min_absolute_delta?: number
      min_expected_count?: number
      cooldown_minutes?: number
      message_template?: string | null
      items_template?: string | null
      message_format?: 'plain' | 'slack_mrkdwn' | 'telegram_html' | 'telegram_markdownv2'
      filters?: AlertRuleFilterPayload[]
    },
  ) => api.patch<AlertRule>(`/projects/${slug}/alert-destinations/${destinationId}/rules/${ruleId}`, data),

  deleteRule: (slug: string, destinationId: string, ruleId: string) =>
    api.del(`/projects/${slug}/alert-destinations/${destinationId}/rules/${ruleId}`),

  simulateRule: (
    slug: string,
    destinationId: string,
    ruleId: string,
    days: number,
    /**
     * Thresholds to try WITHOUT saving them to the rule. Each omitted override
     * falls back to the rule's stored value; the response echoes both as
     * `*_used` / `*_saved` so the preview can be labelled honestly.
     */
    overrides?: {
      cooldownMinutes?: number
      minPercentDelta?: number
      minExpectedCount?: number
      sigmaThreshold?: number
    },
  ) => {
    const params = new URLSearchParams({ days: String(days) })
    if (overrides?.cooldownMinutes !== undefined) {
      params.set('cooldown_minutes_override', String(overrides.cooldownMinutes))
    }
    if (overrides?.minPercentDelta !== undefined) {
      params.set('min_percent_delta_override', String(overrides.minPercentDelta))
    }
    if (overrides?.minExpectedCount !== undefined) {
      params.set('min_expected_count_override', String(overrides.minExpectedCount))
    }
    if (overrides?.sigmaThreshold !== undefined) {
      params.set('sigma_threshold_override', String(overrides.sigmaThreshold))
    }
    return api.post<AlertRuleSimulateResponse>(
      `/projects/${slug}/alert-destinations/${destinationId}/rules/${ruleId}/simulate?${params}`,
      undefined,
    )
  },

  listDeliveries: (
    slug: string,
    params?: {
      status?: string
      channel?: string
      destination_id?: string
      rule_id?: string
      scan_config_id?: string
      /** One incident's deliveries — what the incident card lists under it. */
      correlation_group_id?: string
      /**
       * Deliveries belonging to no incident. Not expressible through
       * `correlation_group_id`, and without it they would vanish from a page
       * that lists deliveries under incidents.
       */
      ungrouped?: boolean
      date_from?: string
      date_to?: string
      offset?: number
      limit?: number
    },
  ) => {
    const sp = new URLSearchParams()
    if (params?.status) sp.set('status', params.status)
    if (params?.channel) sp.set('channel', params.channel)
    if (params?.destination_id) sp.set('destination_id', params.destination_id)
    if (params?.rule_id) sp.set('rule_id', params.rule_id)
    if (params?.scan_config_id) sp.set('scan_config_id', params.scan_config_id)
    if (params?.correlation_group_id) sp.set('correlation_group_id', params.correlation_group_id)
    if (params?.ungrouped) sp.set('ungrouped', 'true')
    if (params?.date_from) sp.set('date_from', params.date_from)
    if (params?.date_to) sp.set('date_to', params.date_to)
    if (params?.offset !== undefined) sp.set('offset', String(params.offset))
    if (params?.limit !== undefined) sp.set('limit', String(params.limit))
    const qs = sp.toString()
    return api.get<AlertDeliveryListResponse>(`/projects/${slug}/alert-deliveries${qs ? `?${qs}` : ''}`)
  },

  getDelivery: (slug: string, deliveryId: string) =>
    api.get<AlertDeliveryDetail>(`/projects/${slug}/alert-deliveries/${deliveryId}`),

  retryDelivery: (slug: string, deliveryId: string) =>
    api.post<AlertDeliveryDetail>(
      `/projects/${slug}/alert-deliveries/${deliveryId}/retry`,
      undefined,
    ),

  getMonitor: (slug: string, ruleId: string) =>
    api.get<MonitorDetail>(`/projects/${slug}/monitors/${ruleId}`),

  getMonitorHistory: (
    slug: string,
    ruleId: string,
    params?: {
      status?: string
      channel?: string
      date_from?: string
      date_to?: string
      offset?: number
      limit?: number
    },
  ) => {
    const sp = new URLSearchParams()
    sp.set('rule_id', ruleId)
    if (params?.status) sp.set('status', params.status)
    if (params?.channel) sp.set('channel', params.channel)
    if (params?.date_from) sp.set('date_from', params.date_from)
    if (params?.date_to) sp.set('date_to', params.date_to)
    if (params?.offset !== undefined) sp.set('offset', String(params.offset))
    if (params?.limit !== undefined) sp.set('limit', String(params.limit))
    return api.get<AlertDeliveryListResponse>(
      `/projects/${slug}/alert-deliveries?${sp.toString()}`,
    )
  },

  muteMonitor: (slug: string, ruleId: string, mutedUntil: string) =>
    api.post<MonitorDetail>(`/projects/${slug}/monitors/${ruleId}/mute`, {
      muted_until: mutedUntil,
    }),

  unmuteMonitor: (slug: string, ruleId: string) =>
    api.post<MonitorDetail>(`/projects/${slug}/monitors/${ruleId}/unmute`, undefined),

  listInbox: (
    slug: string,
    // `status` is the union, not a bare string: the API 422s on anything else,
    // and a filter chip typo should fail at build time rather than as an empty
    // inbox nobody can explain.
    params?: { status?: AlertInboxStatus; offset?: number; limit?: number },
  ) => {
    const sp = new URLSearchParams()
    if (params?.status) sp.set('status', params.status)
    if (params?.offset !== undefined) sp.set('offset', String(params.offset))
    if (params?.limit !== undefined) sp.set('limit', String(params.limit))
    const qs = sp.toString()
    return api.get<AlertInboxListResponse>(`/projects/${slug}/alert-inbox${qs ? `?${qs}` : ''}`)
  },

  /**
   * One incident by id, ignoring the 30-day window the list is scoped to. A
   * deep link to an older incident used to land on an empty inbox because the
   * only way to fetch a group was to find it in that page (tripl-oxkt.11).
   */
  getInboxGroup: (slug: string, correlationGroupId: string) =>
    api.get<AlertInboxGroup>(`/projects/${slug}/alert-inbox/${correlationGroupId}`),

  applyInboxAction: (
    slug: string,
    correlationGroupId: string,
    data: {
      action: AlertInboxAction
      note?: string | null
      muted_until?: string | null
    },
  ) =>
    api.post<AlertInboxActionResponse>(
      `/projects/${slug}/alert-inbox/${correlationGroupId}/actions`,
      data,
    ),

  /**
   * One triage decision, applied to several incidents in ONE request
   * (tripl-gpfr).
   *
   * A triage shortcut, not an incident record: the server copies this body into
   * each selected incident's own state, so afterwards every selected row
   * carries the same note, `acted_at` and `acted_by` and is indistinguishable
   * from N single-incident clicks. There is no group-of-groups object to fetch
   * afterwards, which is why the response hands back the rebuilt cards.
   *
   * ALL-OR-NOTHING. Every id is validated in one query up front; if any is
   * unknown to this project the whole batch is rejected with a 404 and NOTHING
   * was mutated. There is no partial success and no per-item error array, so a
   * caller never has to reconcile "which of my twelve went through".
   *
   * `action` is {@link AlertInboxBulkAction}, so `false_positive` cannot even be
   * spelled here — see that type for why the server refuses it with a 422.
   *
   * `muted_until` follows the single-incident route exactly: pass it only with
   * a `mute`, and pass it EXPLICITLY as `null` for the open-ended mute rather
   * than omitting the key (tripl-a50u). Every other action nulls the column
   * server-side regardless of what is sent.
   */
  applyInboxBulkAction: (
    slug: string,
    data: {
      // First, matching the `<entity>_ids`-first convention of every other bulk
      // body in the repo — and matching the server schema field order, so the
      // two read side by side. Capped at {@link MAX_BULK_INBOX_ACTION_GROUPS} by
      // the server (the only bulk id list in the repo that is capped); a longer
      // list is a 422, not a slow request.
      correlation_group_ids: string[]
      action: AlertInboxBulkAction
      note?: string | null
      muted_until?: string | null
    },
  ) =>
    api.post<AlertInboxBulkActionResponse>(
      `/projects/${slug}/alert-inbox/bulk-actions`,
      data,
    ),
}
